libs/generic-view/utils/src/lib/data-provider-helpers/data-provider-sort.ts

```diff
@@ -4,55 +4,79 @@
  */
 
 import { DataProviderSortConfig } from "device/models"
-import { stringToRegex } from "./string-to-regex"
-import { cloneDeep, get } from "lodash"
-
+import isEmpty from "lodash/isEmpty"
+import {
+  compareFields,
+  compareWithOrderingPatterns,
+  getFirstNonEmptyField,
+  sortByPriority,
+} from "./data-provider-sort.helpers"
 export const dataProviderSort = (
   data: Record<string, unknown>[] = [],
-  sort?: DataProviderSortConfig
+  sortConfigs?: DataProviderSortConfig
 ) => {
-  if (!sort || !data) return data
-  const fieldsSortedByPriority = cloneDeep(sort).sort(
-    (a, b) => a.priority - b.priority
-  )
+  if (!sortConfigs || !data) return data
+
+  const sortedConfigs = sortByPriority(sortConfigs)
 
   return data.sort((a, b) => {
-    let score = 0
     for (const {
       providerField,
+      providerGroup,
       direction,
       orderingPatterns = [],
-    } of fieldsSortedByPriority) {
-      const fieldA = get(a, providerField) as string
-      const fieldB = get(b, providerField) as string
-      if (!fieldA || !fieldB) {
-        continue
-      }
+      sensitivity = "variant",
+      emptyOrder = "last",
+    } of sortedConfigs) {
+      const fields = providerField ? [providerField] : providerGroup
+      if (!fields) continue
+      let index = 0
 
-      for (let i = 0; i < orderingPatterns.length; i++) {
-        const regex = stringToRegex(orderingPatterns[i])
-        const matchA = regex.test(fieldA)
-        const matchB = regex.test(fieldB)
+      while (index < fields.length) {
+        const remainingFields = fields.slice(index)
+        const fieldA = getFirstNonEmptyField(a, remainingFields)
+        const fieldB = getFirstNonEmptyField(b, remainingFields)
 
-        if (matchA && !matchB) {
-          score = -1
-          break
+        if (fieldA === fieldB) {
+          index++
+          continue
         }
-        if (!matchA && matchB) {
-          score = 1
-          break
+
+        if (isEmpty(fieldA) && !isEmpty(fieldB)) {
+          return emptyOrder === "first" ? -1 : 1
+        } else if (!isEmpty(fieldA) && isEmpty(fieldB)) {
+          return emptyOrder === "first" ? 1 : -1
         }
-      }
-      if (score === 0) {
-        score =
-          direction === "asc"
-            ? fieldA.localeCompare(fieldB)
-            : fieldB.localeCompare(fieldA)
-        if (score !== 0) {
-          break
+
+        if (typeof fieldA !== "string" || typeof fieldB !== "string") {
+          index++
+          continue
+        }
```

Comment:
Tests aren't covering this lines. In fact, you can remove them and tests will pass anyway. I'd consider removing it or adding more test cases that will handle such scenario.

Reply:
https://github.com/mudita/mudita-center/pull/2116/commits/d806f4322bb0d1e769f9b001ee281253b943ad07